docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+provisioned-iops volumes, respectively.  From this, it follows that:
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
+used before additional resources are allocated.  In this regard, the existing
+logic need not change much when storage classes are added.  The primary
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
```

Comment:
ACK. Single label can evolve in best practices.

Just want to add, Pod & Replication controllers are user objects. He has full knowledge of it and what to expect using labels.
Storage class is an admin object and users use them. It will be confusing to the user, unless admin spends time creating a description about each label and making it available to end user.